pm: Remove waiting in idle thread when switching to PM_STATE_SOFT_OFF

r2r0 opened the pull request:
The present STM32WB implementation of switching to PM_STATE_SOFT_OFF
uses synchronous call for sending HCI command.
This causes triggering of assertion because waiting is not allowed
in the idle thread.

Fixes #41944

Signed-off-by: Artur Lipowski <Artur.Lipowski@hidglobal.com>

r2r0 committed:
pm: Remove waiting in idle thread when switching to PM_STATE_SOFT_OFF

The present STM32WB implementation of switching to PM_STATE_SOFT_OFF
uses synchronous call for sending HCI command.
This causes triggering of assertion because waiting is not allowed
in the idle thread.

Fixes #41944

Signed-off-by: Artur Lipowski <Artur.Lipowski@hidglobal.com>

erwango commented:
LGTM, but I'd like to be able to test it using dedicated test which is currently not working right now (cf https://github.com/zephyrproject-rtos/zephyr/issues/42308)

erwango commented:
Tested ok with https://github.com/zephyrproject-rtos/zephyr/pull/42274. Removing DNM

r2r0 commented:
After some investigation I think this approach will not work.
- This code is called from pm_system_suspend context with scheduler disabled, so HCI task will have no chance to run to send HCI command.
- There are a few mutex lockings and fifo_get operations (potentail waits) before HCI message is passed to the low level layer.

erwango commented:
> * pm_system_suspend

Thanks for the heads up, would you mind to have a check on `samples/boards/stm32/power_mgmt/stm32wb_ble` and confirm the issue?

r2r0 commented:
I applied #42274 and I added to the `samples/boards/stm32/power_mgmt/stm32wb_ble` a small logging (via printk) if SHUTDOWN is really entrered.
After flashing everthings looks ok but ...
When a cold reset (power off/on) is applied then there is a reset after `Device shutdown` message and applications is restarted.
Then I added:
```
CONFIG_ASSERT=y
CONFIG_ASSERT_LEVEL=2
CONFIG_ASSERT_VERBOSE=y
```
The result (after cold reset) is that MCU does not enter SHUTDOWN (`pm_power_state_set` returns).
I also temporarily increased idle thread stack and I also enabled LOGS to be sure if other problems do not appear but log is clear.

erwango commented:
@r2r0, FYI I've managed to get use case working implementing a simple but proper synchronous `bt_disable`. I'll keep you posted on the progress.

erwango commented:
@r2r0, can you have a check to https://github.com/zephyrproject-rtos/zephyr/pull/42409 and see if solving issues you're seeing ?
